### PR TITLE
017create entity user

### PR DESCRIPTION
**Pull request by OlgaEvdokimova**
Entity User, TechUser, UserRole

### CODE REVIEW

File: src/test/java/by/exadel/internship/mapper/SimpleFormMapperTest.java [hunk 1]
```diff
@@ -1,34 +1,34 @@
...
-        formFullDTO.setEmail("testEmail");
-        formFullDTO.setEducation("testEducation");
-        Form form = mapper.toFormEntity(formFullDTO);
-
-        assertEquals(formFullDTO.getEducation(), form.getEducation());
-        assertEquals(formFullDTO.getEmail(), form.getEmail());
-    }
-}
\ No newline at end of file
+//public class SimpleFormMapperTest {
```
Comment (Fenydude) [inline src/test/java/by/exadel/internship/mapper/SimpleFormMapperTest.java]: It's too much?

**Review by Fenydude — DISMISSED** [review]

File: src/main/java/by/exadel/internship/dto/InterviewDTO.java [hunk 2]
```diff
@@ -1,10 +1,9 @@
 package by.exadel.internship.dto;
-
+import by.exadel.internship.dto.userDTO.UserDTO;
```
Comment (mortal-gpt) [inline src/main/java/by/exadel/internship/dto/InterviewDTO.java]: extra import

File: src/main/java/by/exadel/internship/dto/formDTO/UserRoleDTO.java [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+import lombok.AllArgsConstructor;
+import lombok.Builder;
+import lombok.Data;
+import lombok.NoArgsConstructor;
+import lombok.experimental.SuperBuilder;
+
+@Data
+@NoArgsConstructor
+@AllArgsConstructor
+public class UserRoleDTO {
```
Comment (mortal-gpt) [inline src/main/java/by/exadel/internship/dto/formDTO/UserRoleDTO.java]: Is this  DTO used anywhere? If no, remove it please.

File: src/main/java/by/exadel/internship/dto/internshipDTO/UserInternshipDTO.java [hunk 4]
```diff
@@ -1,7 +1,7 @@
 package by.exadel.internship.dto.internshipDTO;
 
 
-import by.exadel.internship.dto.UserDTO;
+import by.exadel.internship.dto.userDTO.UserDTO;
```
Comment (mortal-gpt) [inline src/main/java/by/exadel/internship/dto/internshipDTO/UserInternshipDTO.java]:
```suggestion
import by.exadel.internship.dto.user.UserDTO;
```

File: src/main/java/by/exadel/internship/dto/userDTO/UserDTO.java [hunk 5]
```diff
@@ -1,18 +1,19 @@
...
 import lombok.NoArgsConstructor;
-
+import lombok.experimental.SuperBuilder;
 import java.time.LocalDateTime;
 import java.util.List;
 import java.util.UUID;
 
 @Data
-@Builder
+@SuperBuilder
```
Comment (mortal-gpt) [inline src/main/java/by/exadel/internship/dto/userDTO/UserDTO.java]: Just remove any builder. It, not needed in an entity

File: src/main/java/by/exadel/internship/entity/user/TechUser.java [hunk 6]
```diff
@@ -0,0 +1,14 @@
+package by.exadel.internship.entity.user;
+import lombok.NoArgsConstructor;
+import lombok.experimental.SuperBuilder;
+import javax.persistence.*;
+
+@Entity
+@Table(name = "techUser")
```
Comment (mortal-gpt) [inline src/main/java/by/exadel/internship/entity/user/TechUser.java]: Will you store tech user in the other table?

File: src/main/java/by/exadel/internship/entity/user/TechUser.java [hunk 7]
```diff
@@ -0,0 +1,14 @@
...
+@NoArgsConstructor
+@SuperBuilder
```
Comment (mortal-gpt) [inline src/main/java/by/exadel/internship/entity/user/TechUser.java]: remove superBuilder

File: src/main/java/by/exadel/internship/entity/user/User.java [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+
+    @Column(name = "u_login")
+    private String login;
+
+    @Column(name = "u_password")
+    private String password;
+
+   @ManyToOne
+   @JoinColumn(name = "u_role")
+    private UserRole userRole;
```
Comment (mortal-gpt) [inline src/main/java/by/exadel/internship/entity/user/User.java:39]: Decided to store it as enum

File: src/main/java/by/exadel/internship/exception/UserNotFoundException.java [hunk 9]
```diff
@@ -0,0 +1,17 @@
+package by.exadel.internship.exception;
+import org.springframework.http.HttpStatus;
+import org.springframework.web.bind.annotation.ResponseStatus;
+
+@ResponseStatus(value = HttpStatus.NOT_FOUND)
+public class UserNotFoundException extends RuntimeException {
```
Comment (mortal-gpt) [inline src/main/java/by/exadel/internship/exception/UserNotFoundException.java]: As far as I understand we want to use not specific exception
```suggestion
public class NotFoundException extends RuntimeException {
```

File: src/main/resources/db/changelog/v.0.1/db.changelog-internship.yaml [hunk 10]
```diff
@@ -18,13 +18,13 @@ databaseChangeLog:
...
-                  type: TIMESTAMP
+                  type: DATE
               - column:
                   name: inship_end_date
-                  type: TIMESTAMP
+                  type: DATE
               - column:
                   name: inship_publication_date
-                  type: TIMESTAMP
+                  type: DATE
```
Comment (mortal-gpt) [inline src/main/resources/db/changelog/v.0.1/db.changelog-internship.yaml:27]: Did you notify other team members about modifying existing changesets?

File: src/main/java/by/exadel/internship/entity/user/User.java [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+import java.util.UUID;
+
+@Entity
+@Table(name = "user")
+@Data
+@AllArgsConstructor
+@NoArgsConstructor
+@SuperBuilder
+@Inheritance(strategy = InheritanceType.SINGLE_TABLE)
+@DiscriminatorColumn(name="u_type",
```
Comment (kzaretskaya) [inline src/main/java/by/exadel/internship/entity/user/User.java]:
```suggestion
@DiscriminatorColumn(name="u_role",
```

File: src/main/resources/application.yaml [hunk 12]
```diff
@@ -2,12 +2,15 @@ spring:
...
+      ddl-auto: validate
   datasource:
     driver-class-name: org.postgresql.Driver
-    url: ${SPRING_DATASOURCE_URL}
-    password: ${SPRING_DATASOURSE_PASSWORD}
-    username: ${SPRING_DATASOURSE_USERNAME}
+#    url: ${SPRING_DATASOURCE_URL}
+#    password: ${SPRING_DATASOURSE_PASSWORD}
+#    username: ${SPRING_DATASOURSE_USERNAME}
+    url: jdbc:postgresql://localhost:5432/internship
```
Comment (kzaretskaya) [inline src/main/resources/application.yaml]: Please, revert these changes to previous state.

File: src/test/java/by/exadel/internship/mapper/SimpleFormMapperTest.java (hunk 1 above)
Comment (OlgaEvdokimova) [reply src/test/java/by/exadel/internship/mapper/SimpleFormMapperTest.java]: done

File: src/main/java/by/exadel/internship/dto/InterviewDTO.java (hunk 2 above)
Comment (OlgaEvdokimova) [reply src/main/java/by/exadel/internship/dto/InterviewDTO.java]: done

File: src/main/java/by/exadel/internship/dto/formDTO/UserRoleDTO.java (hunk 3 above)
Comment (OlgaEvdokimova) [reply src/main/java/by/exadel/internship/dto/formDTO/UserRoleDTO.java]: done

File: src/main/java/by/exadel/internship/entity/user/TechUser.java (hunk 6 above)
Comment (OlgaEvdokimova) [reply src/main/java/by/exadel/internship/entity/user/TechUser.java]: done

File: src/main/java/by/exadel/internship/entity/user/TechUser.java (hunk 7 above)
Comment (OlgaEvdokimova) [reply src/main/java/by/exadel/internship/entity/user/TechUser.java]: done

File: src/main/java/by/exadel/internship/entity/user/User.java (hunk 8 above)
Comment (OlgaEvdokimova) [reply src/main/java/by/exadel/internship/entity/user/User.java:39]: done

File: src/main/java/by/exadel/internship/exception/UserNotFoundException.java (hunk 9 above)
Comment (OlgaEvdokimova) [reply src/main/java/by/exadel/internship/exception/UserNotFoundException.java]: done

File: src/main/resources/db/changelog/v.0.1/db.changelog-internship.yaml (hunk 10 above)
Comment (OlgaEvdokimova) [reply src/main/resources/db/changelog/v.0.1/db.changelog-internship.yaml:27]: yes i have told them

File: src/main/java/by/exadel/internship/entity/user/User.java (hunk 11 above)
Comment (OlgaEvdokimova) [reply src/main/java/by/exadel/internship/entity/user/User.java]: We decided not to use inheritance

File: src/main/resources/application.yaml (hunk 12 above)
Comment (OlgaEvdokimova) [reply src/main/resources/application.yaml]: done

**Review by velichko1882 — DISMISSED** [review]

File: src/main/java/by/exadel/internship/entity/user/TechUser.java [hunk 13]
```diff
@@ -0,0 +1,9 @@
+package by.exadel.internship.entity.user;
+import lombok.NoArgsConstructor;
+
+
+@NoArgsConstructor
+public class TechUser  {
```
Comment (mortal-gpt) [inline src/main/java/by/exadel/internship/entity/user/TechUser.java]: As far as I remember we decided to remove this class.

File: src/main/java/by/exadel/internship/dto/user/TechUserDTO.java [hunk 14]
```diff
@@ -0,0 +1,13 @@
+package by.exadel.internship.dto.user;
+import by.exadel.internship.dto.enums.Technology;
+import lombok.AllArgsConstructor;
+import lombok.Data;
+import lombok.NoArgsConstructor;
+
+@Data
+@AllArgsConstructor
+@NoArgsConstructor
+public class TechUserDTO extends UserDTO {
```
Comment (mortal-gpt) [inline src/main/java/by/exadel/internship/dto/user/TechUserDTO.java]: As far as I remember we decided to remove this class.

File: src/main/java/by/exadel/internship/mapper/TechUserMapper.java [hunk 15]
```diff
@@ -0,0 +1,19 @@
+package by.exadel.internship.mapper;
+import by.exadel.internship.dto.user.TechUserDTO;
+import by.exadel.internship.entity.user.TechUser;
+import org.mapstruct.IterableMapping;
+import org.mapstruct.Mapper;
+import org.mapstruct.Named;
+import java.util.List;
+
+@Mapper(componentModel = "spring")
+public interface TechUserMapper {
```
Comment (mortal-gpt) [inline src/main/java/by/exadel/internship/mapper/TechUserMapper.java]: As far as I remember we decided to remove this class.

File: src/main/java/by/exadel/internship/entity/user/User.java [hunk 16]
```diff
@@ -0,0 +1,48 @@
...
+import javax.persistence.*;
+import java.util.List;
+import java.util.UUID;
+
+@Entity
+@Table(name = "user")
+@Data
+@AllArgsConstructor
+@NoArgsConstructor
+
```
Comment (kzaretskaya) [inline src/main/java/by/exadel/internship/entity/user/User.java:16]: Minor: remove empty line

File: src/main/resources/db/changelog/v.0.2/db.changelog-user.yaml [hunk 17]
```diff
@@ -36,19 +36,23 @@ databaseChangeLog:
...
       changes:
-        - addForeignKeyConstraint:
-            baseColumnNames: u_role_id
-            baseTableName: user
-            constraintName: fk_user_role
-            referencedColumnNames: r_id
-            referencedTableName: role
+        - addColumn:
+            tableName: user
+            columns:
```
Comment (mortal-gpt) [inline src/main/resources/db/changelog/v.0.2/db.changelog-user.yaml]: move this column into "createTable" section

File: src/main/java/by/exadel/internship/dto/internshipDTO/UserInternshipDTO.java (hunk 4 above)
Comment (OlgaEvdokimova) [reply src/main/java/by/exadel/internship/dto/internshipDTO/UserInternshipDTO.java]: done

File: src/main/java/by/exadel/internship/dto/userDTO/UserDTO.java (hunk 5 above)
Comment (OlgaEvdokimova) [reply src/main/java/by/exadel/internship/dto/userDTO/UserDTO.java]: done

File: src/main/resources/db/changelog/v.0.2/db.changelog-user.yaml (hunk 17 above)
Comment (OlgaEvdokimova) [reply src/main/resources/db/changelog/v.0.2/db.changelog-user.yaml]: done

File: src/main/java/by/exadel/internship/entity/user/User.java (hunk 16 above)
Comment (OlgaEvdokimova) [reply src/main/java/by/exadel/internship/entity/user/User.java:16]: done

File: src/main/java/by/exadel/internship/mapper/TechUserMapper.java (hunk 15 above)
Comment (OlgaEvdokimova) [reply src/main/java/by/exadel/internship/mapper/TechUserMapper.java]: Sorry for my misunderstanding,  I've dropped it. Done

File: src/main/java/by/exadel/internship/entity/user/TechUser.java (hunk 13 above)
Comment (OlgaEvdokimova) [reply src/main/java/by/exadel/internship/entity/user/TechUser.java]: Sorry for my misunderstanding, I've dropped it. Done

File: src/main/java/by/exadel/internship/dto/user/TechUserDTO.java (hunk 14 above)
Comment (OlgaEvdokimova) [reply src/main/java/by/exadel/internship/dto/user/TechUserDTO.java]: done

**Review by velichko1882 — DISMISSED** [review]

**Review by kzaretskaya — DISMISSED** [review]

**Review by kzaretskaya — APPROVED** [review]

**Review by velichko1882 — APPROVED** [review]

**Review by viktarkolbik — APPROVED** [review]

File: src/main/java/by/exadel/internship/entity/user/User.java (hunk 11 above)
Comment (OlgaEvdokimova) [reply src/main/java/by/exadel/internship/entity/user/User.java]: done
